File: liquidator/src/main/cpp/sor/src/order_splitter.cpp
#include "microoptimus/sor/order_splitter.hpp"
#include <algorithm>

namespace microoptimus {
namespace sor {

std::vector<VenueAllocation> OrderSplitter::splitOrder(
    const OrderRequest& order,
    const std::vector<VenueType>& venues,
    const VenueScorer& scorer) const {

    std::vector<VenueAllocation> allocations;

    if (venues.empty()) {
        return allocations;
    }

    // Single venue - no splitting needed
    if (venues.size() == 1) {
        allocations.emplace_back(venues[0], order.quantity, 1);
        return allocations;
    }

    // VWAP-style allocation: prioritize venues by score with capacity limits
    int64_t remainingQuantity = order.quantity;
    int priority = 1;

    for (const auto& venue : venues) {
        if (remainingQuantity <= 0) break;

        const VenueConfig* config = scorer.getVenueConfig(venue);
        if (!config || !config->enabled) continue;

        // Allocate min of: remaining quantity, venue max capacity, or proportional share
        int64_t venueCapacity = std::min(config->maxOrderSize, remainingQuantity);

        // VWAP-style proportional allocation
        int64_t allocation;
        if (priority == 1) {
            // Best venue gets 40% allocation
            allocation = std::min(venueCapacity, (remainingQuantity * 40) / 100);
        } else if (priority == 2) {
            // Second venue gets 30% allocation
            allocation = std::min(venueCapacity, (remainingQuantity * 30) / 100);
        } else {
            // Remaining venues split the rest proportionally
            int64_t remainingVenues = venues.size() - priority + 1;
            allocation = std::min(venueCapacity, remainingQuantity / remainingVenues);
        }

        // Ensure we allocate something if there's remaining quantity
        if (allocation == 0 && remainingQuantity > 0) {
            allocation = std::min(venueCapacity, remainingQuantity);
        }

        if (allocation > 0) {
            allocations.emplace_back(venue, allocation, priority);
            remainingQuantity -= allocation;
            priority++;
        }
    }

    // If still have remaining quantity, add to last allocation
    if (remainingQuantity > 0 && !allocations.empty()) {
        allocations.back().quantity += remainingQuantity;
    }

    return allocations;
}

} // namespace sor
} // namespace microoptimus

File: osm/src/main/java/com/microoptimus/osm/sor/OrderSplitter.java
package com.microoptimus.osm.sor;

import java.util.Arrays;

/**
 * OrderSplitter - Splits large orders across multiple venues
 *
 * Strategies:
 * - Pro-rata: Split proportionally by venue score
 * - Sequential: Fill best venue first, then next
 * - Balanced: Equal split across top venues
 */
public class OrderSplitter {

    public enum SplitStrategy {
        PRO_RATA,       // Proportional to venue scores
        SEQUENTIAL,     // Best venue first
        BALANCED        // Equal split
    }

    private SplitStrategy strategy = SplitStrategy.PRO_RATA;
    private int maxVenues = 3;  // Maximum venues to split across

    // Pre-allocated allocations (GC-free)
    private final VenueAllocation[] allocations;

    public OrderSplitter() {
        this.allocations = new VenueAllocation[VenueId.values().length];
        for (int i = 0; i < allocations.length; i++) {
            allocations[i] = new VenueAllocation();
        }
    }

    /**
     * Split order across venues based on scores
     */
    public VenueAllocation[] splitOrder(OrderRequest request, VenueScore[] scores,
                                         VenueConfig[] configs) {
        // Sort scores by total score (descending)
        VenueScore[] sortedScores = sortByScore(scores);

        // Count valid venues
        int validVenues = 0;
        for (VenueScore score : sortedScores) {
            if (score != null && score.totalScore > 0) {
                validVenues++;
            }
        }

        if (validVenues == 0) {
            return new VenueAllocation[0];
        }

        int venuesToUse = Math.min(validVenues, maxVenues);
        VenueAllocation[] result = new VenueAllocation[venuesToUse];

        switch (strategy) {
            case PRO_RATA:
                splitProRata(result, request, sortedScores, venuesToUse);
                break;
            case SEQUENTIAL:
                splitSequential(result, request, sortedScores, configs, venuesToUse);
                break;
            case BALANCED:
                splitBalanced(result, request, sortedScores, venuesToUse);
                break;
        }

        return result;
    }

    /**
     * Pro-rata split based on venue scores
     */
    private void splitProRata(VenueAllocation[] result, OrderRequest request,
                               VenueScore[] sortedScores, int venuesToUse) {
        // Calculate total score
        double totalScore = 0;
        for (int i = 0; i < venuesToUse; i++) {
            totalScore += sortedScores[i].totalScore;
        }

        // Allocate proportionally
        long remaining = request.quantity;
        for (int i = 0; i < venuesToUse; i++) {
            VenueScore score = sortedScores[i];
            double proportion = score.totalScore / totalScore;
            long qty;

            if (i == venuesToUse - 1) {
                // Last venue gets remainder
                qty = remaining;
            } else {
                qty = Math.min(remaining, (long) (request.quantity * proportion));
                qty = Math.min(qty, score.maxQuantity);
            }

            result[i] = new VenueAllocation();
            result[i].init(score.venueId, qty, i + 1,
                    score.estimatedLatencyNanos,
                    score.estimatedFillProbability,
                    score.estimatedCost * qty / request.quantity);

            remaining -= qty;
        }
    }

    /**
     * Sequential split - fill best venue up to max, then next
     */
    private void splitSequential(VenueAllocation[] result, OrderRequest request,
                                  VenueScore[] sortedScores, VenueConfig[] configs,
                                  int venuesToUse) {
        long remaining = request.quantity;

        for (int i = 0; i < venuesToUse && remaining > 0; i++) {
            VenueScore score = sortedScores[i];
            long maxQty = score.maxQuantity;
            long qty = Math.min(remaining, maxQty);

            result[i] = new VenueAllocation();
            result[i].init(score.venueId, qty, i + 1,
                    score.estimatedLatencyNanos,
                    score.estimatedFillProbability,
                    score.estimatedCost * qty / request.quantity);

            remaining -= qty;
        }
    }

    /**
     * Balanced split - equal allocation across venues
     */
    private void splitBalanced(VenueAllocation[] result, OrderRequest request,
                                VenueScore[] sortedScores, int venuesToUse) {
        long baseQty = request.quantity / venuesToUse;
        long remainder = request.quantity % venuesToUse;

        for (int i = 0; i < venuesToUse; i++) {
            VenueScore score = sortedScores[i];
            long qty = baseQty + (i < remainder ? 1 : 0);
            qty = Math.min(qty, score.maxQuantity);

            result[i] = new VenueAllocation();
            result[i].init(score.venueId, qty, i + 1,
                    score.estimatedLatencyNanos,
                    score.estimatedFillProbability,
                    score.estimatedCost * qty / request.quantity);
        }
    }

    /**
     * Sort scores by total score (descending)
     */
    private VenueScore[] sortByScore(VenueScore[] scores) {
        VenueScore[] sorted = Arrays.copyOf(scores, scores.length);
        Arrays.sort(sorted, (a, b) -> {
            if (a == null || a.totalScore == 0) return 1;
            if (b == null || b.totalScore == 0) return -1;
            return Double.compare(b.totalScore, a.totalScore);
        });
        return sorted;
    }

    // Configuration
    public void setStrategy(SplitStrategy strategy) { this.strategy = strategy; }
    public void setMaxVenues(int max) { this.maxVenues = max; }
    public SplitStrategy getStrategy() { return strategy; }
    public int getMaxVenues() { return maxVenues; }
}
